Fix key generation

artygus opened the pull request:
Query string keys must be escaped after the split bacause:
- Query string could look like 'weird%3Dkey=whatever'
- Rake util's unescape method forces utf-8 encoding on strings, which will cause the "invalid byte sequence in UTF-8" error for the most string methods called on badly encoded strings, e.g. Rack::Utils::unescape('%D0%BA=%D1').split('=')

grosser commented:
can you add a test for that ?

I'd prefer `map { |s| unescape(s) }` to be more normal / less fancy

grosser commented:
can it use something more standard like `@request.params` ?

artygus commented:
@grosser I added tests and rebased the commit. I guess we can't use `params` because `params` are merged with POST, although we could use @request.GET, but in that case resulting hash may have other hashes inside, e.g. from query string like this `/test?a[b]=c&a[d]=e`, and before Ruby 1.9 hashes didn't preserve keys ordering, so something like recursive sorting must implemented

grosser commented:
thx for fixing this nasty bug :)
